{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5016962a-8e5a-4494-9c90-7d7065164bf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4fd179cf-b21c-4bcf-a213-40541da473e1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Global seed set to 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Random seed set to 42\n",
      "scanpy              : 1.9.3\n",
      "anndata             : 0.8.0\n",
      "scvi                : 0.20.3\n",
      "hiddensc            : beta_25.03.23\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "root_path = os.path.abspath('./..')\n",
    "sys.path.insert(0, root_path )\n",
    "\n",
    "import itertools\n",
    "import functools\n",
    "from tqdm import tqdm\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import numba\n",
    "import sklearn\n",
    "import sklearn.linear_model\n",
    "import sklearn.cluster\n",
    "import sklearn.metrics\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import hiddensc\n",
    "from hiddensc import utils, files, vis\n",
    "import seaborn as sns\n",
    "\n",
    "import scanpy as sc\n",
    "import scvi\n",
    "import anndata\n",
    "#from . import mixscape\n",
    "\n",
    "utils.set_random_seed(utils.RANDOM_SEED)\n",
    "utils.print_module_versions([sc, anndata, scvi, hiddensc])\n",
    "vis.visual_settings()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "50db456a-1be1-4c30-b3a8-313e8f4beb36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"2\" halign=\"left\">PCA</th>\n",
       "      <th>batch</th>\n",
       "      <th>perturbed</th>\n",
       "      <th colspan=\"4\" halign=\"left\">PCA</th>\n",
       "      <th colspan=\"2\" halign=\"left\">LinearSCVI_10</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"2\" halign=\"left\">SCVI_10</th>\n",
       "      <th colspan=\"4\" halign=\"left\">SCVI_25</th>\n",
       "      <th colspan=\"4\" halign=\"left\">SCVI_50</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"2\" halign=\"left\">Mixscape</th>\n",
       "      <th>Unnamed: 1_level_1</th>\n",
       "      <th>Unnamed: 2_level_1</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Logistic</th>\n",
       "      <th colspan=\"2\" halign=\"left\">SVM</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Logistic</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"2\" halign=\"left\">SVM</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Logistic</th>\n",
       "      <th colspan=\"2\" halign=\"left\">SVM</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Logistic</th>\n",
       "      <th colspan=\"2\" halign=\"left\">SVM</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>p_hat</th>\n",
       "      <th>p_label</th>\n",
       "      <th>Case</th>\n",
       "      <th>Memory B</th>\n",
       "      <th>p_hat</th>\n",
       "      <th>p_label</th>\n",
       "      <th>p_hat</th>\n",
       "      <th>p_label</th>\n",
       "      <th>p_hat</th>\n",
       "      <th>p_label</th>\n",
       "      <th>...</th>\n",
       "      <th>p_hat</th>\n",
       "      <th>p_label</th>\n",
       "      <th>p_hat</th>\n",
       "      <th>p_label</th>\n",
       "      <th>p_hat</th>\n",
       "      <th>p_label</th>\n",
       "      <th>p_hat</th>\n",
       "      <th>p_label</th>\n",
       "      <th>p_hat</th>\n",
       "      <th>p_label</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcodes</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>naiveB_a_AAACCTGCACGGTAGA-1</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.564</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.568</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.490</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.458</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.487</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.464</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.490</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>naiveB_a_AAACCTGCAGATGGGT-1</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.406</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.442</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.538</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.424</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.467</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.490</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.469</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.491</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>naiveB_a_AAAGATGCATTTCAGG-1</th>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.369</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.439</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.411</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.483</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.466</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.491</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.363</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.460</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>naiveB_a_AAAGCAAAGCCAACAG-1</th>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.496</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.494</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.442</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.463</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.493</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.400</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.475</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>naiveB_a_AAAGCAAAGTGCCATT-1</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.489</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.488</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.502</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.459</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.469</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.490</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.543</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.515</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>memoryB_b_TGTCCCATCCTTTACA-1</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.688</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.568</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.689</td>\n",
       "      <td>1.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.468</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.515</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.463</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.486</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>memoryB_b_TTGACTTAGCTTTGGT-1</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.927</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.725</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.740</td>\n",
       "      <td>1.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.609</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.581</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.516</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.708</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.538</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>memoryB_b_TTGCCGTAGGCGTACA-1</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.792</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.599</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.745</td>\n",
       "      <td>1.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.550</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.649</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.532</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.630</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.526</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>memoryB_b_TTTGCGCTCTACTATC-1</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.703</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.567</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.696</td>\n",
       "      <td>1.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.641</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.631</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.527</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.658</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.538</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>memoryB_b_TTTGGTTGTTTGTGTG-1</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.783</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.616</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.723</td>\n",
       "      <td>1.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.591</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.675</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.539</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.743</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.563</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2000 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  PCA                      batch  \\\n",
       "                             Mixscape         Unnamed: 1_level_1   \n",
       "                                p_hat p_label               Case   \n",
       "barcodes                                                           \n",
       "naiveB_a_AAACCTGCACGGTAGA-1     0.000   0.000              1.000   \n",
       "naiveB_a_AAACCTGCAGATGGGT-1     0.000   0.000              1.000   \n",
       "naiveB_a_AAAGATGCATTTCAGG-1     0.000   1.000              0.000   \n",
       "naiveB_a_AAAGCAAAGCCAACAG-1     0.000   1.000              0.000   \n",
       "naiveB_a_AAAGCAAAGTGCCATT-1     0.000   0.000              1.000   \n",
       "...                               ...     ...                ...   \n",
       "memoryB_b_TGTCCCATCCTTTACA-1    0.000   0.000              1.000   \n",
       "memoryB_b_TTGACTTAGCTTTGGT-1    0.000   0.000              1.000   \n",
       "memoryB_b_TTGCCGTAGGCGTACA-1    0.000   0.000              1.000   \n",
       "memoryB_b_TTTGCGCTCTACTATC-1    0.000   0.000              1.000   \n",
       "memoryB_b_TTTGGTTGTTTGTGTG-1    0.000   0.000              1.000   \n",
       "\n",
       "                                      perturbed      PCA                \\\n",
       "                             Unnamed: 2_level_1 Logistic           SVM   \n",
       "                                       Memory B    p_hat p_label p_hat   \n",
       "barcodes                                                                 \n",
       "naiveB_a_AAACCTGCACGGTAGA-1               0.000    0.564   1.000 0.568   \n",
       "naiveB_a_AAACCTGCAGATGGGT-1               0.000    0.406   0.000 0.442   \n",
       "naiveB_a_AAAGATGCATTTCAGG-1               0.000    0.369   0.000 0.439   \n",
       "naiveB_a_AAAGCAAAGCCAACAG-1               0.000    0.496   0.000 0.494   \n",
       "naiveB_a_AAAGCAAAGTGCCATT-1               0.000    0.489   0.000 0.488   \n",
       "...                                         ...      ...     ...   ...   \n",
       "memoryB_b_TGTCCCATCCTTTACA-1              1.000    0.688   1.000 0.568   \n",
       "memoryB_b_TTGACTTAGCTTTGGT-1              1.000    0.927   1.000 0.725   \n",
       "memoryB_b_TTGCCGTAGGCGTACA-1              1.000    0.792   1.000 0.599   \n",
       "memoryB_b_TTTGCGCTCTACTATC-1              1.000    0.703   1.000 0.567   \n",
       "memoryB_b_TTTGGTTGTTTGTGTG-1              1.000    0.783   1.000 0.616   \n",
       "\n",
       "                                     LinearSCVI_10          ... SCVI_10  \\\n",
       "                                          Logistic          ...     SVM   \n",
       "                             p_label         p_hat p_label  ...   p_hat   \n",
       "barcodes                                                    ...           \n",
       "naiveB_a_AAACCTGCACGGTAGA-1    1.000         0.490   0.000  ...   0.458   \n",
       "naiveB_a_AAACCTGCAGATGGGT-1    0.000         0.538   0.000  ...   0.424   \n",
       "naiveB_a_AAAGATGCATTTCAGG-1    0.000         0.411   0.000  ...   0.483   \n",
       "naiveB_a_AAAGCAAAGCCAACAG-1    0.000         0.442   0.000  ...   0.500   \n",
       "naiveB_a_AAAGCAAAGTGCCATT-1    0.000         0.502   0.000  ...   0.459   \n",
       "...                              ...           ...     ...  ...     ...   \n",
       "memoryB_b_TGTCCCATCCTTTACA-1   1.000         0.689   1.000  ...   0.468   \n",
       "memoryB_b_TTGACTTAGCTTTGGT-1   1.000         0.740   1.000  ...   0.609   \n",
       "memoryB_b_TTGCCGTAGGCGTACA-1   1.000         0.745   1.000  ...   0.550   \n",
       "memoryB_b_TTTGCGCTCTACTATC-1   1.000         0.696   1.000  ...   0.641   \n",
       "memoryB_b_TTTGGTTGTTTGTGTG-1   1.000         0.723   1.000  ...   0.591   \n",
       "\n",
       "                                      SCVI_25                        SCVI_50  \\\n",
       "                                     Logistic           SVM         Logistic   \n",
       "                             p_label    p_hat p_label p_hat p_label    p_hat   \n",
       "barcodes                                                                       \n",
       "naiveB_a_AAACCTGCACGGTAGA-1    0.000    0.487   0.000 0.500   0.000    0.464   \n",
       "naiveB_a_AAACCTGCAGATGGGT-1    0.000    0.467   0.000 0.490   0.000    0.469   \n",
       "naiveB_a_AAAGATGCATTTCAGG-1    0.000    0.466   0.000 0.491   0.000    0.363   \n",
       "naiveB_a_AAAGCAAAGCCAACAG-1    0.000    0.463   0.000 0.493   0.000    0.400   \n",
       "naiveB_a_AAAGCAAAGTGCCATT-1    0.000    0.469   0.000 0.490   0.000    0.543   \n",
       "...                              ...      ...     ...   ...     ...      ...   \n",
       "memoryB_b_TGTCCCATCCTTTACA-1   0.000    0.515   0.000 0.500   0.000    0.463   \n",
       "memoryB_b_TTGACTTAGCTTTGGT-1   1.000    0.581   1.000 0.516   1.000    0.708   \n",
       "memoryB_b_TTGCCGTAGGCGTACA-1   1.000    0.649   1.000 0.532   1.000    0.630   \n",
       "memoryB_b_TTTGCGCTCTACTATC-1   1.000    0.631   1.000 0.527   1.000    0.658   \n",
       "memoryB_b_TTTGGTTGTTTGTGTG-1   1.000    0.675   1.000 0.539   1.000    0.743   \n",
       "\n",
       "                                                    \n",
       "                                       SVM          \n",
       "                             p_label p_hat p_label  \n",
       "barcodes                                            \n",
       "naiveB_a_AAACCTGCACGGTAGA-1    0.000 0.490   0.000  \n",
       "naiveB_a_AAACCTGCAGATGGGT-1    0.000 0.491   0.000  \n",
       "naiveB_a_AAAGATGCATTTCAGG-1    0.000 0.460   0.000  \n",
       "naiveB_a_AAAGCAAAGCCAACAG-1    0.000 0.475   0.000  \n",
       "naiveB_a_AAAGCAAAGTGCCATT-1    1.000 0.515   1.000  \n",
       "...                              ...   ...     ...  \n",
       "memoryB_b_TGTCCCATCCTTTACA-1   0.000 0.486   0.000  \n",
       "memoryB_b_TTGACTTAGCTTTGGT-1   1.000 0.538   1.000  \n",
       "memoryB_b_TTGCCGTAGGCGTACA-1   1.000 0.526   1.000  \n",
       "memoryB_b_TTTGCGCTCTACTATC-1   1.000 0.538   1.000  \n",
       "memoryB_b_TTTGGTTGTTTGTGTG-1   1.000 0.563   1.000  \n",
       "\n",
       "[2000 rows x 32 columns]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import glob\n",
    "\n",
    "pred_files = list(glob.glob('../results/naiveB_1900_memoryB_100/*predictions.csv'))\n",
    "\n",
    "dirname = '../results/naiveB_1900_memoryB_100/'\n",
    "\n",
    "def _read_prediction_csv(fname:str) -> pd.DataFrame:\n",
    "    \"\"\"Predictions have a special header + index, so we have a helper function.\"\"\"\n",
    "    files.check_ext(fname, 'csv')\n",
    "    return pd.read_csv(fname, index_col=[0], header=[0, 1, 2])\n",
    "\n",
    "\n",
    "dfs = [_read_prediction_csv(fname) for fname in glob.glob(os.path.join(dirname, '*predictions.csv'))]\n",
    "pred_df = pd.concat(dfs, axis=1)\n",
    "pred_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "0007868d-ee8e-4dda-b41f-f9d7c9769e9c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'batch': array([1., 1., 0., ..., 1., 1., 1.]),\n",
       " 'perturbed': array([0., 0., 0., ..., 1., 1., 1.])}"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs = [_read_prediction_csv(fname) for fname in glob.glob(os.path.join(dirname, '*predictions.csv'))]\n",
    "pred_df = pd.concat(dfs, axis=1)\n",
    "cols = pred_df.columns.tolist()\n",
    "values = {}\n",
    "for label in ['batch', 'perturbed']:\n",
    "    is_col = np.array([label in c[0] for c in cols])\n",
    "    assert np.sum(is_col) == 1, f'Expected at least one {label} column, found {np.sum(is_col)}'\n",
    "    index = np.argmax(is_col)\n",
    "    values[label] = pred_df[cols[index]].values\n",
    "    pred_df.drop(cols[index], axis=1, inplace=True)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "9bf54a4b-0c92-4d9e-8f84-939daf619e11",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "      <th>perturbed</th>\n",
       "      <th colspan=\"4\" halign=\"left\">PCA</th>\n",
       "      <th colspan=\"4\" halign=\"left\">LinearSCVI_10</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"2\" halign=\"left\">SCVI_10</th>\n",
       "      <th colspan=\"4\" halign=\"left\">SCVI_25</th>\n",
       "      <th colspan=\"4\" halign=\"left\">SCVI_50</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>Unnamed: 1_level_1</th>\n",
       "      <th>Unnamed: 2_level_1</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Logistic</th>\n",
       "      <th colspan=\"2\" halign=\"left\">SVM</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Logistic</th>\n",
       "      <th colspan=\"2\" halign=\"left\">SVM</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"2\" halign=\"left\">SVM</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Logistic</th>\n",
       "      <th colspan=\"2\" halign=\"left\">SVM</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Logistic</th>\n",
       "      <th colspan=\"2\" halign=\"left\">SVM</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>Case</th>\n",
       "      <th>Memory B</th>\n",
       "      <th>p_hat</th>\n",
       "      <th>p_label</th>\n",
       "      <th>p_hat</th>\n",
       "      <th>p_label</th>\n",
       "      <th>p_hat</th>\n",
       "      <th>p_label</th>\n",
       "      <th>p_hat</th>\n",
       "      <th>p_label</th>\n",
       "      <th>...</th>\n",
       "      <th>p_hat</th>\n",
       "      <th>p_label</th>\n",
       "      <th>p_hat</th>\n",
       "      <th>p_label</th>\n",
       "      <th>p_hat</th>\n",
       "      <th>p_label</th>\n",
       "      <th>p_hat</th>\n",
       "      <th>p_label</th>\n",
       "      <th>p_hat</th>\n",
       "      <th>p_label</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcodes</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>naiveB_a_AAACCTGCACGGTAGA-1</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.564</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.568</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.490</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.520</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.458</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.487</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.464</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.490</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>naiveB_a_AAACCTGCAGATGGGT-1</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.406</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.442</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.538</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.537</td>\n",
       "      <td>1.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.424</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.467</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.490</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.469</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.491</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>naiveB_a_AAAGATGCATTTCAGG-1</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.369</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.439</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.411</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.447</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.483</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.466</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.491</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.363</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.460</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>naiveB_a_AAAGCAAAGCCAACAG-1</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.496</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.494</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.442</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.452</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.463</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.493</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.400</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.475</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>naiveB_a_AAAGCAAAGTGCCATT-1</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.489</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.488</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.502</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.495</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.459</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.469</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.490</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.543</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.515</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>memoryB_b_TGTCCCATCCTTTACA-1</th>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.688</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.568</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.689</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.609</td>\n",
       "      <td>1.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.468</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.515</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.463</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.486</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>memoryB_b_TTGACTTAGCTTTGGT-1</th>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.927</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.725</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.740</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.631</td>\n",
       "      <td>1.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.609</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.581</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.516</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.708</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.538</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>memoryB_b_TTGCCGTAGGCGTACA-1</th>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.792</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.599</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.745</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.651</td>\n",
       "      <td>1.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.550</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.649</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.532</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.630</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.526</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>memoryB_b_TTTGCGCTCTACTATC-1</th>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.703</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.567</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.696</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.609</td>\n",
       "      <td>1.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.641</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.631</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.527</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.658</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.538</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>memoryB_b_TTTGGTTGTTTGTGTG-1</th>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.783</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.616</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.723</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.630</td>\n",
       "      <td>1.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.591</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.675</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.539</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.743</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.563</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2000 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          batch          perturbed      PCA  \\\n",
       "                             Unnamed: 1_level_1 Unnamed: 2_level_1 Logistic   \n",
       "                                           Case           Memory B    p_hat   \n",
       "barcodes                                                                      \n",
       "naiveB_a_AAACCTGCACGGTAGA-1               1.000              0.000    0.564   \n",
       "naiveB_a_AAACCTGCAGATGGGT-1               1.000              0.000    0.406   \n",
       "naiveB_a_AAAGATGCATTTCAGG-1               0.000              0.000    0.369   \n",
       "naiveB_a_AAAGCAAAGCCAACAG-1               0.000              0.000    0.496   \n",
       "naiveB_a_AAAGCAAAGTGCCATT-1               1.000              0.000    0.489   \n",
       "...                                         ...                ...      ...   \n",
       "memoryB_b_TGTCCCATCCTTTACA-1              1.000              1.000    0.688   \n",
       "memoryB_b_TTGACTTAGCTTTGGT-1              1.000              1.000    0.927   \n",
       "memoryB_b_TTGCCGTAGGCGTACA-1              1.000              1.000    0.792   \n",
       "memoryB_b_TTTGCGCTCTACTATC-1              1.000              1.000    0.703   \n",
       "memoryB_b_TTTGGTTGTTTGTGTG-1              1.000              1.000    0.783   \n",
       "\n",
       "                                                   LinearSCVI_10          \\\n",
       "                                       SVM              Logistic           \n",
       "                             p_label p_hat p_label         p_hat p_label   \n",
       "barcodes                                                                   \n",
       "naiveB_a_AAACCTGCACGGTAGA-1    1.000 0.568   1.000         0.490   0.000   \n",
       "naiveB_a_AAACCTGCAGATGGGT-1    0.000 0.442   0.000         0.538   0.000   \n",
       "naiveB_a_AAAGATGCATTTCAGG-1    0.000 0.439   0.000         0.411   0.000   \n",
       "naiveB_a_AAAGCAAAGCCAACAG-1    0.000 0.494   0.000         0.442   0.000   \n",
       "naiveB_a_AAAGCAAAGTGCCATT-1    0.000 0.488   0.000         0.502   0.000   \n",
       "...                              ...   ...     ...           ...     ...   \n",
       "memoryB_b_TGTCCCATCCTTTACA-1   1.000 0.568   1.000         0.689   1.000   \n",
       "memoryB_b_TTGACTTAGCTTTGGT-1   1.000 0.725   1.000         0.740   1.000   \n",
       "memoryB_b_TTGCCGTAGGCGTACA-1   1.000 0.599   1.000         0.745   1.000   \n",
       "memoryB_b_TTTGCGCTCTACTATC-1   1.000 0.567   1.000         0.696   1.000   \n",
       "memoryB_b_TTTGGTTGTTTGTGTG-1   1.000 0.616   1.000         0.723   1.000   \n",
       "\n",
       "                                            ... SCVI_10          SCVI_25  \\\n",
       "                               SVM          ...     SVM         Logistic   \n",
       "                             p_hat p_label  ...   p_hat p_label    p_hat   \n",
       "barcodes                                    ...                            \n",
       "naiveB_a_AAACCTGCACGGTAGA-1  0.520   0.000  ...   0.458   0.000    0.487   \n",
       "naiveB_a_AAACCTGCAGATGGGT-1  0.537   1.000  ...   0.424   0.000    0.467   \n",
       "naiveB_a_AAAGATGCATTTCAGG-1  0.447   0.000  ...   0.483   0.000    0.466   \n",
       "naiveB_a_AAAGCAAAGCCAACAG-1  0.452   0.000  ...   0.500   0.000    0.463   \n",
       "naiveB_a_AAAGCAAAGTGCCATT-1  0.495   0.000  ...   0.459   0.000    0.469   \n",
       "...                            ...     ...  ...     ...     ...      ...   \n",
       "memoryB_b_TGTCCCATCCTTTACA-1 0.609   1.000  ...   0.468   0.000    0.515   \n",
       "memoryB_b_TTGACTTAGCTTTGGT-1 0.631   1.000  ...   0.609   1.000    0.581   \n",
       "memoryB_b_TTGCCGTAGGCGTACA-1 0.651   1.000  ...   0.550   1.000    0.649   \n",
       "memoryB_b_TTTGCGCTCTACTATC-1 0.609   1.000  ...   0.641   1.000    0.631   \n",
       "memoryB_b_TTTGGTTGTTTGTGTG-1 0.630   1.000  ...   0.591   1.000    0.675   \n",
       "\n",
       "                                                    SCVI_50                \\\n",
       "                                       SVM         Logistic           SVM   \n",
       "                             p_label p_hat p_label    p_hat p_label p_hat   \n",
       "barcodes                                                                    \n",
       "naiveB_a_AAACCTGCACGGTAGA-1    0.000 0.500   0.000    0.464   0.000 0.490   \n",
       "naiveB_a_AAACCTGCAGATGGGT-1    0.000 0.490   0.000    0.469   0.000 0.491   \n",
       "naiveB_a_AAAGATGCATTTCAGG-1    0.000 0.491   0.000    0.363   0.000 0.460   \n",
       "naiveB_a_AAAGCAAAGCCAACAG-1    0.000 0.493   0.000    0.400   0.000 0.475   \n",
       "naiveB_a_AAAGCAAAGTGCCATT-1    0.000 0.490   0.000    0.543   1.000 0.515   \n",
       "...                              ...   ...     ...      ...     ...   ...   \n",
       "memoryB_b_TGTCCCATCCTTTACA-1   0.000 0.500   0.000    0.463   0.000 0.486   \n",
       "memoryB_b_TTGACTTAGCTTTGGT-1   1.000 0.516   1.000    0.708   1.000 0.538   \n",
       "memoryB_b_TTGCCGTAGGCGTACA-1   1.000 0.532   1.000    0.630   1.000 0.526   \n",
       "memoryB_b_TTTGCGCTCTACTATC-1   1.000 0.527   1.000    0.658   1.000 0.538   \n",
       "memoryB_b_TTTGGTTGTTTGTGTG-1   1.000 0.539   1.000    0.743   1.000 0.563   \n",
       "\n",
       "                                      \n",
       "                                      \n",
       "                             p_label  \n",
       "barcodes                              \n",
       "naiveB_a_AAACCTGCACGGTAGA-1    0.000  \n",
       "naiveB_a_AAACCTGCAGATGGGT-1    0.000  \n",
       "naiveB_a_AAAGATGCATTTCAGG-1    0.000  \n",
       "naiveB_a_AAAGCAAAGCCAACAG-1    0.000  \n",
       "naiveB_a_AAAGCAAAGTGCCATT-1    1.000  \n",
       "...                              ...  \n",
       "memoryB_b_TGTCCCATCCTTTACA-1   0.000  \n",
       "memoryB_b_TTGACTTAGCTTTGGT-1   1.000  \n",
       "memoryB_b_TTGCCGTAGGCGTACA-1   1.000  \n",
       "memoryB_b_TTTGCGCTCTACTATC-1   1.000  \n",
       "memoryB_b_TTTGGTTGTTTGTGTG-1   1.000  \n",
       "\n",
       "[2000 rows x 30 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fname = '../results/naiveB_1900_memoryB_100/predictions.csv'\n",
    "pred_df = pd.read_csv(fname, index_col=[0], header=[0, 1, 2])\n",
    "a_df = pd.read_csv('../results/naiveB_1900_memoryB_100/mixscape_predictions.csv', index_col=[0], header=[0, 1, 2])\n",
    "pred_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "41e08290-62ae-4580-9b32-a778350eeac6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "      <th>perturbed</th>\n",
       "      <th colspan=\"4\" halign=\"left\">PCA</th>\n",
       "      <th colspan=\"4\" halign=\"left\">LinearSCVI_10</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"4\" halign=\"left\">SCVI_25</th>\n",
       "      <th colspan=\"4\" halign=\"left\">SCVI_50</th>\n",
       "      <th colspan=\"2\" halign=\"left\">PCA</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>Unnamed: 1_level_1</th>\n",
       "      <th>Unnamed: 2_level_1</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Logistic</th>\n",
       "      <th colspan=\"2\" halign=\"left\">SVM</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Logistic</th>\n",
       "      <th colspan=\"2\" halign=\"left\">SVM</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Logistic</th>\n",
       "      <th colspan=\"2\" halign=\"left\">SVM</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Logistic</th>\n",
       "      <th colspan=\"2\" halign=\"left\">SVM</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Mixscape</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>Case</th>\n",
       "      <th>Memory B</th>\n",
       "      <th>p_hat</th>\n",
       "      <th>p_label</th>\n",
       "      <th>p_hat</th>\n",
       "      <th>p_label</th>\n",
       "      <th>p_hat</th>\n",
       "      <th>p_label</th>\n",
       "      <th>p_hat</th>\n",
       "      <th>p_label</th>\n",
       "      <th>...</th>\n",
       "      <th>p_hat</th>\n",
       "      <th>p_label</th>\n",
       "      <th>p_hat</th>\n",
       "      <th>p_label</th>\n",
       "      <th>p_hat</th>\n",
       "      <th>p_label</th>\n",
       "      <th>p_hat</th>\n",
       "      <th>p_label</th>\n",
       "      <th>p_hat</th>\n",
       "      <th>p_label</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcodes</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>naiveB_a_AAACCTGCACGGTAGA-1</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.564</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.568</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.490</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.520</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.487</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.464</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.490</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>naiveB_a_AAACCTGCAGATGGGT-1</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.406</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.442</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.538</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.537</td>\n",
       "      <td>1.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.467</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.490</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.469</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.491</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>naiveB_a_AAAGATGCATTTCAGG-1</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.369</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.439</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.411</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.447</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.466</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.491</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.363</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.460</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>naiveB_a_AAAGCAAAGCCAACAG-1</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.496</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.494</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.442</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.452</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.463</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.493</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.400</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.475</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>naiveB_a_AAAGCAAAGTGCCATT-1</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.489</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.488</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.502</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.495</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.469</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.490</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.543</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.515</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>memoryB_b_TGTCCCATCCTTTACA-1</th>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.688</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.568</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.689</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.609</td>\n",
       "      <td>1.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.515</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.463</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.486</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>memoryB_b_TTGACTTAGCTTTGGT-1</th>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.927</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.725</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.740</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.631</td>\n",
       "      <td>1.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.581</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.516</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.708</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.538</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>memoryB_b_TTGCCGTAGGCGTACA-1</th>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.792</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.599</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.745</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.651</td>\n",
       "      <td>1.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.649</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.532</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.630</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.526</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>memoryB_b_TTTGCGCTCTACTATC-1</th>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.703</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.567</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.696</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.609</td>\n",
       "      <td>1.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.631</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.527</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.658</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.538</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>memoryB_b_TTTGGTTGTTTGTGTG-1</th>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.783</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.616</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.723</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.630</td>\n",
       "      <td>1.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.675</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.539</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.743</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.563</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2000 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          batch          perturbed      PCA  \\\n",
       "                             Unnamed: 1_level_1 Unnamed: 2_level_1 Logistic   \n",
       "                                           Case           Memory B    p_hat   \n",
       "barcodes                                                                      \n",
       "naiveB_a_AAACCTGCACGGTAGA-1               1.000              0.000    0.564   \n",
       "naiveB_a_AAACCTGCAGATGGGT-1               1.000              0.000    0.406   \n",
       "naiveB_a_AAAGATGCATTTCAGG-1               0.000              0.000    0.369   \n",
       "naiveB_a_AAAGCAAAGCCAACAG-1               0.000              0.000    0.496   \n",
       "naiveB_a_AAAGCAAAGTGCCATT-1               1.000              0.000    0.489   \n",
       "...                                         ...                ...      ...   \n",
       "memoryB_b_TGTCCCATCCTTTACA-1              1.000              1.000    0.688   \n",
       "memoryB_b_TTGACTTAGCTTTGGT-1              1.000              1.000    0.927   \n",
       "memoryB_b_TTGCCGTAGGCGTACA-1              1.000              1.000    0.792   \n",
       "memoryB_b_TTTGCGCTCTACTATC-1              1.000              1.000    0.703   \n",
       "memoryB_b_TTTGGTTGTTTGTGTG-1              1.000              1.000    0.783   \n",
       "\n",
       "                                                   LinearSCVI_10          \\\n",
       "                                       SVM              Logistic           \n",
       "                             p_label p_hat p_label         p_hat p_label   \n",
       "barcodes                                                                   \n",
       "naiveB_a_AAACCTGCACGGTAGA-1    1.000 0.568   1.000         0.490   0.000   \n",
       "naiveB_a_AAACCTGCAGATGGGT-1    0.000 0.442   0.000         0.538   0.000   \n",
       "naiveB_a_AAAGATGCATTTCAGG-1    0.000 0.439   0.000         0.411   0.000   \n",
       "naiveB_a_AAAGCAAAGCCAACAG-1    0.000 0.494   0.000         0.442   0.000   \n",
       "naiveB_a_AAAGCAAAGTGCCATT-1    0.000 0.488   0.000         0.502   0.000   \n",
       "...                              ...   ...     ...           ...     ...   \n",
       "memoryB_b_TGTCCCATCCTTTACA-1   1.000 0.568   1.000         0.689   1.000   \n",
       "memoryB_b_TTGACTTAGCTTTGGT-1   1.000 0.725   1.000         0.740   1.000   \n",
       "memoryB_b_TTGCCGTAGGCGTACA-1   1.000 0.599   1.000         0.745   1.000   \n",
       "memoryB_b_TTTGCGCTCTACTATC-1   1.000 0.567   1.000         0.696   1.000   \n",
       "memoryB_b_TTTGGTTGTTTGTGTG-1   1.000 0.616   1.000         0.723   1.000   \n",
       "\n",
       "                                            ...  SCVI_25                \\\n",
       "                               SVM          ... Logistic           SVM   \n",
       "                             p_hat p_label  ...    p_hat p_label p_hat   \n",
       "barcodes                                    ...                          \n",
       "naiveB_a_AAACCTGCACGGTAGA-1  0.520   0.000  ...    0.487   0.000 0.500   \n",
       "naiveB_a_AAACCTGCAGATGGGT-1  0.537   1.000  ...    0.467   0.000 0.490   \n",
       "naiveB_a_AAAGATGCATTTCAGG-1  0.447   0.000  ...    0.466   0.000 0.491   \n",
       "naiveB_a_AAAGCAAAGCCAACAG-1  0.452   0.000  ...    0.463   0.000 0.493   \n",
       "naiveB_a_AAAGCAAAGTGCCATT-1  0.495   0.000  ...    0.469   0.000 0.490   \n",
       "...                            ...     ...  ...      ...     ...   ...   \n",
       "memoryB_b_TGTCCCATCCTTTACA-1 0.609   1.000  ...    0.515   0.000 0.500   \n",
       "memoryB_b_TTGACTTAGCTTTGGT-1 0.631   1.000  ...    0.581   1.000 0.516   \n",
       "memoryB_b_TTGCCGTAGGCGTACA-1 0.651   1.000  ...    0.649   1.000 0.532   \n",
       "memoryB_b_TTTGCGCTCTACTATC-1 0.609   1.000  ...    0.631   1.000 0.527   \n",
       "memoryB_b_TTTGGTTGTTTGTGTG-1 0.630   1.000  ...    0.675   1.000 0.539   \n",
       "\n",
       "                                      SCVI_50                            PCA  \\\n",
       "                                     Logistic           SVM         Mixscape   \n",
       "                             p_label    p_hat p_label p_hat p_label    p_hat   \n",
       "barcodes                                                                       \n",
       "naiveB_a_AAACCTGCACGGTAGA-1    0.000    0.464   0.000 0.490   0.000    0.000   \n",
       "naiveB_a_AAACCTGCAGATGGGT-1    0.000    0.469   0.000 0.491   0.000    0.000   \n",
       "naiveB_a_AAAGATGCATTTCAGG-1    0.000    0.363   0.000 0.460   0.000    0.000   \n",
       "naiveB_a_AAAGCAAAGCCAACAG-1    0.000    0.400   0.000 0.475   0.000    0.000   \n",
       "naiveB_a_AAAGCAAAGTGCCATT-1    0.000    0.543   1.000 0.515   1.000    0.000   \n",
       "...                              ...      ...     ...   ...     ...      ...   \n",
       "memoryB_b_TGTCCCATCCTTTACA-1   0.000    0.463   0.000 0.486   0.000    0.000   \n",
       "memoryB_b_TTGACTTAGCTTTGGT-1   1.000    0.708   1.000 0.538   1.000    0.000   \n",
       "memoryB_b_TTGCCGTAGGCGTACA-1   1.000    0.630   1.000 0.526   1.000    0.000   \n",
       "memoryB_b_TTTGCGCTCTACTATC-1   1.000    0.658   1.000 0.538   1.000    0.000   \n",
       "memoryB_b_TTTGGTTGTTTGTGTG-1   1.000    0.743   1.000 0.563   1.000    0.000   \n",
       "\n",
       "                                      \n",
       "                                      \n",
       "                             p_label  \n",
       "barcodes                              \n",
       "naiveB_a_AAACCTGCACGGTAGA-1    0.000  \n",
       "naiveB_a_AAACCTGCAGATGGGT-1    0.000  \n",
       "naiveB_a_AAAGATGCATTTCAGG-1    1.000  \n",
       "naiveB_a_AAAGCAAAGCCAACAG-1    1.000  \n",
       "naiveB_a_AAAGCAAAGTGCCATT-1    0.000  \n",
       "...                              ...  \n",
       "memoryB_b_TGTCCCATCCTTTACA-1   0.000  \n",
       "memoryB_b_TTGACTTAGCTTTGGT-1   0.000  \n",
       "memoryB_b_TTGCCGTAGGCGTACA-1   0.000  \n",
       "memoryB_b_TTTGCGCTCTACTATC-1   0.000  \n",
       "memoryB_b_TTTGGTTGTTTGTGTG-1   0.000  \n",
       "\n",
       "[2000 rows x 32 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "b_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "979149dd-1fc8-4328-9ac1-19809022f089",
   "metadata": {},
   "outputs": [],
   "source": [
    "b_df = pd.concat((pred_df, a_df), axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38251957-2f11-4d8a-a330-d1d4735f2291",
   "metadata": {},
   "source": [
    "# Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9be9fe7b-78df-45b8-bc89-439829f63cf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "OVERWRITE = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "4c182847-8a27-4b3f-873d-0914e2461e16",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Random seed set to 42\n"
     ]
    }
   ],
   "source": [
    "utils.set_random_seed(utils.RANDOM_SEED)\n",
    "exp_id = files.MEMORY_B_VALUES[0]\n",
    "# Check and setup files.\n",
    "data_name = f'naiveB_1900_memoryB_{exp_id:d}'\n",
    "at_results_dir = functools.partial(os.path.join, root_path, files.RESULT_DIR, data_name)\n",
    "if not os.path.exists(at_results_dir('dim_reduced.npz')):\n",
    "    print(f'Did not find features file for {data_name}')\n",
    "\n",
    "if os.path.exists(at_results_dir('predictions.csv')) and not OVERWRITE:\n",
    "    print(f'Skipping {data_name}, found outputs and overwrite={OVERWRITE}')\n",
    "\n",
    "# Load dataset.\n",
    "main_datafile = os.path.join(root_path, files.DATA_DIR, f'{data_name}_raw.h5ad')\n",
    "adata = sc.read(main_datafile)\n",
    "adata.X = adata.layers['counts']\n",
    "sc.pp.normalize_total(adata)\n",
    "sc.pp.log1p(adata)\n",
    "sc.pp.highly_variable_genes(adata, subset=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "a1ff1dfe-6038-49fe-ae60-11b20bf4be87",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Pertpy's implementation of Mixscape.\n",
    "\n",
    "We have this custom module since the current github repo version is 0.4 but it breaks.\n",
    "Pip is 0.3 but is not as flexible as wee need for our comparison.\n",
    "Source reference: https://github.com/theislab/pertpy/blob/development/pertpy/tools/_mixscape.py\n",
    "Pertpy's liscence is MIT (https://github.com/theislab/pertpy/blob/development/LICENSE)\n",
    "\"\"\"\n",
    "\n",
    "from __future__ import annotations\n",
    "\n",
    "import warnings\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "from anndata import AnnData\n",
    "from pynndescent import NNDescent\n",
    "from rich import print\n",
    "from scanpy.tools._utils import _choose_representation\n",
    "from scipy import sparse\n",
    "from scipy.sparse import csr_matrix, issparse, spmatrix\n",
    "from sklearn.mixture import GaussianMixture\n",
    "\n",
    "\n",
    "warnings.simplefilter(\"ignore\")\n",
    "\n",
    "\n",
    "class Mixscape:\n",
    "    \"\"\"Python implementation of Mixscape.\"\"\"\n",
    "\n",
    "    def __init__(self):\n",
    "        pass\n",
    "\n",
    "    def perturbation_signature(\n",
    "        self,\n",
    "        adata: AnnData,\n",
    "        pert_key: str,\n",
    "        control: str,\n",
    "        split_by: str | None = None,\n",
    "        n_neighbors: int = 20,\n",
    "        use_rep: str | None = None,\n",
    "        n_pcs: int | None = None,\n",
    "        batch_size: int | None = None,\n",
    "        copy: bool = False,\n",
    "        **kwargs,\n",
    "    ):\n",
    "        \"\"\"Calculate perturbation signature.\n",
    "\n",
    "        For each cell, we identify `n_neighbors` cells from the control pool with the most similar mRNA expression profiles.\n",
    "        The perturbation signature is calculated by subtracting the averaged mRNA expression profile of the control\n",
    "        neighbors from the mRNA expression profile of each cell.\n",
    "\n",
    "        Args:\n",
    "            adata: The annotated data object.\n",
    "            pert_key: The column  of `.obs` with perturbation categories, should also contain `control`.\n",
    "            control: Control category from the `pert_key` column.\n",
    "            split_by: Provide the column `.obs` if multiple biological replicates exist to calculate\n",
    "                the perturbation signature for every replicate separately.\n",
    "            n_neighbors: Number of neighbors from the control to use for the perturbation signature.\n",
    "            use_rep: Use the indicated representation. `'X'` or any key for `.obsm` is valid.\n",
    "                If `None`, the representation is chosen automatically:\n",
    "                For `.n_vars` < 50, `.X` is used, otherwise 'X_pca' is used.\n",
    "                If 'X_pca' is not present, it’s computed with default parameters.\n",
    "            n_pcs: Use this many PCs. If `n_pcs==0` use `.X` if `use_rep is None`.\n",
    "            batch_size: Size of batch to calculate the perturbation signature.\n",
    "                If 'None', the perturbation signature is calcuated in the full mode, requiring more memory.\n",
    "                The batched mode is very inefficient for sparse data.\n",
    "            copy: Determines whether a copy of the `adata` is returned.\n",
    "            **kwargs: Additional arguments for the `NNDescent` class from `pynndescent`.\n",
    "\n",
    "        Returns:\n",
    "            If `copy=True`, returns the copy of `adata` with the perturbation signature in `.layers[\"X_pert\"]`.\n",
    "            Otherwise writes the perturbation signature directly to `.layers[\"X_pert\"]` of the provided `adata`.\n",
    "        \"\"\"\n",
    "        if copy:\n",
    "            adata = adata.copy()\n",
    "\n",
    "        adata.layers[\"X_pert\"] = adata.X.copy()\n",
    "\n",
    "        control_mask = adata.obs[pert_key] == control\n",
    "\n",
    "        if split_by is None:\n",
    "            split_masks = [np.full(adata.n_obs, True, dtype=bool)]\n",
    "        else:\n",
    "            split_obs = adata.obs[split_by]\n",
    "            cats = split_obs.unique()\n",
    "            split_masks = [split_obs == cat for cat in cats]\n",
    "\n",
    "        R = _choose_representation(adata, use_rep=use_rep, n_pcs=n_pcs)\n",
    "\n",
    "        for split_mask in split_masks:\n",
    "            control_mask_split = control_mask & split_mask\n",
    "\n",
    "            R_split = R[split_mask]\n",
    "            R_control = R[control_mask_split]\n",
    "\n",
    "            eps = kwargs.pop(\"epsilon\", 0.1)\n",
    "            nn_index = NNDescent(R_control, **kwargs)\n",
    "            indices, _ = nn_index.query(R_split, k=n_neighbors, epsilon=eps)\n",
    "\n",
    "            X_control = np.expm1(adata.X[control_mask_split])\n",
    "\n",
    "            n_split = split_mask.sum()\n",
    "            n_control = X_control.shape[0]\n",
    "\n",
    "            if batch_size is None:\n",
    "                col_indices = np.ravel(indices)\n",
    "                row_indices = np.repeat(np.arange(n_split), n_neighbors)\n",
    "\n",
    "                neigh_matrix = csr_matrix(\n",
    "                    (np.ones_like(col_indices, dtype=np.float64), (row_indices, col_indices)),\n",
    "                    shape=(n_split, n_control),\n",
    "                )\n",
    "                neigh_matrix /= n_neighbors\n",
    "                adata.layers[\"X_pert\"][split_mask] -= np.log1p(neigh_matrix @ X_control)\n",
    "            else:\n",
    "                is_sparse = issparse(X_control)\n",
    "                split_indices = np.where(split_mask)[0]\n",
    "                for i in range(0, n_split, batch_size):\n",
    "                    size = min(i + batch_size, n_split)\n",
    "                    select = slice(i, size)\n",
    "\n",
    "                    batch = np.ravel(indices[select])\n",
    "                    split_batch = split_indices[select]\n",
    "\n",
    "                    size = size - i\n",
    "\n",
    "                    # sparse is very slow\n",
    "                    means_batch = X_control[batch]\n",
    "                    means_batch = means_batch.toarray() if is_sparse else means_batch\n",
    "                    means_batch = means_batch.reshape(size, n_neighbors, -1).mean(1)\n",
    "\n",
    "                    adata.layers[\"X_pert\"][split_batch] -= np.log1p(means_batch)\n",
    "\n",
    "        if copy:\n",
    "            return adata\n",
    "\n",
    "    def mixscape(\n",
    "        self,\n",
    "        adata: AnnData,\n",
    "        labels: str,\n",
    "        control: str,\n",
    "        new_class_name: str | None = \"mixscape_class\",\n",
    "        min_de_genes: int | None = 5,\n",
    "        layer: str | None = None,\n",
    "        logfc_threshold: float | None = 0.25,\n",
    "        iter_num: int | None = 10,\n",
    "        split_by: str | None = None,\n",
    "        pval_cutoff: float | None = 5e-2,\n",
    "        perturbation_type: str | None = \"KO\",\n",
    "        copy: bool | None = False,\n",
    "    ):\n",
    "        \"\"\"Identify perturbed and non-perturbed gRNA expressing cells that accounts for multiple treatments/conditions/chemical perturbations.\n",
    "\n",
    "        The implementation resembles https://satijalab.org/seurat/reference/runmixscape\n",
    "\n",
    "        Args:\n",
    "            adata: The annotated data object.\n",
    "            pert_key: The column of `.obs` with perturbation categories, should also contain `control`.\n",
    "            labels: The column of `.obs` with target gene labels.\n",
    "            control: Control category from the `pert_key` column.\n",
    "            new_class_name: Name of mixscape classification to be stored in `.obs`.\n",
    "            min_de_genes: Required number of genes that are differentially expressed for method to separate perturbed and non-perturbed cells.\n",
    "            layer: Key from adata.layers whose value will be used to perform tests on. Default is using `.layers[\"X_pert\"]`.\n",
    "            logfc_threshold: Limit testing to genes which show, on average, at least X-fold difference (log-scale) between the two groups of cells (default: 0.25).\n",
    "            iter_num: Number of normalmixEM iterations to run if convergence does not occur.\n",
    "            split_by: Provide the column `.obs` if multiple biological replicates exist to calculate\n",
    "                    the perturbation signature for every replicate separately.\n",
    "            pval_cutoff: P-value cut-off for selection of significantly DE genes.\n",
    "            perturbation_type: specify type of CRISPR perturbation expected for labeling mixscape classifications. Defaults to KO.\n",
    "            copy: Determines whether a copy of the `adata` is returned.\n",
    "\n",
    "        Returns:\n",
    "            If `copy=True`, returns the copy of `adata` with the classification result in `.obs`.\n",
    "            Otherwise writes the results directly to `.obs` of the provided `adata`.\n",
    "\n",
    "            mixscape_class: pandas.Series (`adata.obs['mixscape_class']`).\n",
    "            Classification result with cells being either classified as perturbed (KO, by default) or non-perturbed (NP) based on their target gene class.\n",
    "\n",
    "            mixscape_class_global: pandas.Series (`adata.obs['mixscape_class_global']`).\n",
    "            Global classification result (perturbed, NP or NT)\n",
    "\n",
    "            mixscape_class_p_ko: pandas.Series (`adata.obs['mixscape_class_p_ko']`).\n",
    "            Posterior probabilities used to determine if a cell is KO (default).\n",
    "            Name of this item will change to match perturbation_type parameter setting. (>0.5) or NP\n",
    "        \"\"\"\n",
    "        if copy:\n",
    "            adata = adata.copy()\n",
    "\n",
    "        if split_by is None:\n",
    "            split_masks = [np.full(adata.n_obs, True, dtype=bool)]\n",
    "            categories = [\"all\"]\n",
    "        else:\n",
    "            split_obs = adata.obs[split_by]\n",
    "            categories = split_obs.unique()\n",
    "            split_masks = [split_obs == category for category in categories]\n",
    "\n",
    "        perturbation_markers = self._get_perturbation_markers(\n",
    "            adata, split_masks, categories, labels, control, layer, pval_cutoff, min_de_genes, logfc_threshold\n",
    "        )\n",
    "\n",
    "        adata_comp = adata\n",
    "        if layer is not None:\n",
    "            X = adata_comp.layers[layer]\n",
    "        else:\n",
    "            try:\n",
    "                X = adata_comp.layers[\"X_pert\"]\n",
    "            except KeyError:\n",
    "                print(\n",
    "                    '[bold yellow]No \"X_pert\" found in .layers! -- Please run pert_sign first to calculate perturbation signature!'\n",
    "                )\n",
    "                raise\n",
    "        # initialize return variables\n",
    "        adata.obs[f\"{new_class_name}_p_{perturbation_type.lower()}\"] = 0\n",
    "        adata.obs[new_class_name] = adata.obs[labels].astype(str)\n",
    "        adata.obs[f\"{new_class_name}_global\"] = np.empty(\n",
    "            [\n",
    "                adata.n_obs,\n",
    "            ],\n",
    "            dtype=np.object_,\n",
    "        )\n",
    "        gv_list: dict[str, dict] = {}\n",
    "        for split, split_mask in enumerate(split_masks):\n",
    "            category = categories[split]\n",
    "            genes = list(set(adata[split_mask].obs[labels]).difference([control]))\n",
    "            for gene in genes:\n",
    "                post_prob = 0\n",
    "                orig_guide_cells = (adata.obs[labels] == gene) & split_mask\n",
    "                orig_guide_cells_index = list(orig_guide_cells.index[orig_guide_cells])\n",
    "                nt_cells = (adata.obs[labels] == control) & split_mask\n",
    "                all_cells = orig_guide_cells | nt_cells\n",
    "                if len(perturbation_markers[(category, gene)]) == 0:\n",
    "                    adata.obs.loc[orig_guide_cells, new_class_name] = f\"{gene} NP\"\n",
    "                    print(f'Found 0 perturbation markers for {category, gene}')\n",
    "                else:\n",
    "                    de_genes = perturbation_markers[(category, gene)]\n",
    "                    de_genes_indices = self._get_column_indices(adata, list(de_genes))\n",
    "                    dat = X[all_cells][:, de_genes_indices]\n",
    "                    converged = False\n",
    "                    n_iter = 0\n",
    "                    old_classes = adata.obs[labels][all_cells]\n",
    "                    while not converged and n_iter < iter_num:\n",
    "                        # Get all cells in current split&Gene\n",
    "                        guide_cells = (adata.obs[labels] == gene) & split_mask\n",
    "                        # get average value for each gene over all selected cells\n",
    "                        # all cells in current split&Gene minus all NT cells in current split\n",
    "                        # Each row is for each cell, each column is for each gene, get mean for each column\n",
    "                        vec = np.mean(X[guide_cells][:, de_genes_indices], axis=0) - np.mean(\n",
    "                            X[nt_cells][:, de_genes_indices], axis=0\n",
    "                        )\n",
    "                        # project cells onto the perturbation vector\n",
    "                        if isinstance(dat, spmatrix):\n",
    "                            pvec = np.sum(np.multiply(dat.toarray(), vec), axis=1) / np.sum(np.multiply(vec, vec))\n",
    "                        else:\n",
    "                            pvec = np.sum(np.multiply(dat, vec), axis=1) / np.sum(np.multiply(vec, vec))\n",
    "                        pvec = pd.Series(np.asarray(pvec).flatten(), index=list(all_cells.index[all_cells]))\n",
    "                        if n_iter == 0:\n",
    "                            gv = pd.DataFrame(columns=[\"pvec\", labels])\n",
    "                            gv[\"pvec\"] = pvec\n",
    "                            gv[labels] = control\n",
    "                            gv.loc[guide_cells, labels] = gene\n",
    "                            if gene not in gv_list.keys():\n",
    "                                gv_list[gene] = {}\n",
    "                            gv_list[gene][category] = gv\n",
    "\n",
    "                        guide_norm = self._define_normal_mixscape(pvec[guide_cells])\n",
    "                        nt_norm = self._define_normal_mixscape(pvec[nt_cells])\n",
    "                        means_init = np.array([[nt_norm[0]], [guide_norm[0]]])\n",
    "                        precisions_init = np.array([nt_norm[1], guide_norm[1]])\n",
    "                        mm = GaussianMixture(\n",
    "                            n_components=2,\n",
    "                            covariance_type=\"spherical\",\n",
    "                            means_init=means_init,\n",
    "                            precisions_init=precisions_init,\n",
    "                        ).fit(np.asarray(pvec).reshape(-1, 1))\n",
    "                        probabilities = mm.predict_proba(np.array(pvec[orig_guide_cells_index]).reshape(-1, 1))\n",
    "                        extra['prob']=probabilities\n",
    "                        lik_ratio = probabilities[:, 0] / probabilities[:, 1]\n",
    "                        post_prob = 1 / (1 + lik_ratio)\n",
    "                        extra['phat'] = post_prob\n",
    "                        extra['labels'] = post_prob > 0.5\n",
    "                        \n",
    "                        # based on the posterior probability, assign cells to the two classes\n",
    "                        adata.obs.loc[\n",
    "                            [orig_guide_cells_index[cell] for cell in np.where(post_prob > 0.5)[0]], new_class_name\n",
    "                        ] = gene\n",
    "                        adata.obs.loc[\n",
    "                            [orig_guide_cells_index[cell] for cell in np.where(post_prob <= 0.5)[0]], new_class_name\n",
    "                        ] = f\"{gene} NP\"\n",
    "                        if sum(adata.obs[new_class_name][split_mask] == gene) < min_de_genes:\n",
    "                            adata.obs.loc[guide_cells, new_class_name] = \"NP\"\n",
    "                            converged = True\n",
    "                        if adata.obs[new_class_name][all_cells].equals(old_classes):\n",
    "                            converged = True\n",
    "                        old_classes = adata.obs[new_class_name][all_cells]\n",
    "                        n_iter += 1\n",
    "\n",
    "                    adata.obs.loc[\n",
    "                        (adata.obs[new_class_name] == gene) & split_mask, new_class_name\n",
    "                    ] = f\"{gene} {perturbation_type}\"\n",
    "\n",
    "                adata.obs[f\"{new_class_name}_global\"] = [a.split(\" \")[-1] for a in adata.obs[new_class_name]]\n",
    "                adata.obs.loc[orig_guide_cells_index, f\"{new_class_name}_p_{perturbation_type.lower()}\"] = post_prob\n",
    "        adata.uns[\"mixscape\"] = gv_list\n",
    "\n",
    "        if copy:\n",
    "            return adata\n",
    "\n",
    "\n",
    "    def _get_perturbation_markers(\n",
    "        self,\n",
    "        adata: AnnData,\n",
    "        split_masks: list[np.ndarray],\n",
    "        categories: list[str],\n",
    "        labels: str,\n",
    "        control: str,\n",
    "        layer: str,\n",
    "        pval_cutoff: float,\n",
    "        min_de_genes: float,\n",
    "        logfc_threshold: float,\n",
    "    ) -> dict[tuple, np.ndarray]:\n",
    "        \"\"\"determine gene sets across all splits/groups through differential gene expression\n",
    "\n",
    "        Args:\n",
    "            adata: :class:`~anndata.AnnData` object\n",
    "            col_names: Column names to extract the indices for\n",
    "\n",
    "        Returns:\n",
    "            Set of column indices.\n",
    "        \"\"\"\n",
    "        perturbation_markers: dict[tuple, np.ndarray] = {}\n",
    "        for split, split_mask in enumerate(split_masks):\n",
    "            category = categories[split]\n",
    "            # get gene sets for each split\n",
    "            genes = list(set(adata[split_mask].obs[labels]).difference([control]))\n",
    "            adata_split = adata[split_mask].copy()\n",
    "            # find top DE genes between cells with targeting and non-targeting gRNAs\n",
    "            sc.tl.rank_genes_groups(\n",
    "                adata_split, layer=layer, groupby=labels, groups=genes, reference=control, method=\"t-test\"\n",
    "            )\n",
    "            # get DE genes for each gene\n",
    "            for gene in genes:\n",
    "                logfc_threshold_mask = adata_split.uns[\"rank_genes_groups\"][\"logfoldchanges\"][gene] >= logfc_threshold\n",
    "                de_genes = adata_split.uns[\"rank_genes_groups\"][\"names\"][gene][logfc_threshold_mask]\n",
    "                pvals_adj = adata_split.uns[\"rank_genes_groups\"][\"pvals_adj\"][gene][logfc_threshold_mask]\n",
    "                de_genes = de_genes[pvals_adj < pval_cutoff]\n",
    "                if len(de_genes) < min_de_genes:\n",
    "                    de_genes = np.array([])\n",
    "                perturbation_markers[(category, gene)] = de_genes\n",
    "\n",
    "        return perturbation_markers\n",
    "\n",
    "    def _get_column_indices(self, adata, col_names):\n",
    "        \"\"\"Fetches the column indices in X for a given list of column names\n",
    "\n",
    "        Args:\n",
    "            adata: :class:`~anndata.AnnData` object\n",
    "            col_names: Column names to extract the indices for\n",
    "\n",
    "        Returns:\n",
    "            Set of column indices\n",
    "        \"\"\"\n",
    "        if isinstance(col_names, str):  # pragma: no cover\n",
    "            col_names = [col_names]\n",
    "\n",
    "        indices = list()\n",
    "        for idx, col in enumerate(adata.var_names):\n",
    "            if col in col_names:\n",
    "                indices.append(idx)\n",
    "\n",
    "        return indices\n",
    "\n",
    "    def _define_normal_mixscape(\n",
    "        self, X: np.ndarray | sparse.spmatrix | pd.DataFrame | None\n",
    "    ) -> list[float]:  # noqa: N803\n",
    "        \"\"\"Calculates the mean and standard deviation of a matrix.\n",
    "\n",
    "        Args:\n",
    "            X: The matrix to calculate the properties for.\n",
    "\n",
    "        Returns:\n",
    "            Mean and standard deviation of the matrix.\n",
    "        \"\"\"\n",
    "        mu = X.mean()\n",
    "        sd = X.std()\n",
    "\n",
    "        return [mu, sd]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "1cac9242-2772-4c8c-a313-8e9bcc0fc063",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">0</span>          Case\n",
       "<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">1</span>          Case\n",
       "<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2</span>       Control\n",
       "<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">3</span>       Control\n",
       "<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">4</span>          Case\n",
       "         <span style=\"color: #808000; text-decoration-color: #808000\">...</span>   \n",
       "<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">1944</span>       Case\n",
       "<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">1945</span>       Case\n",
       "<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">1946</span>       Case\n",
       "<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">1947</span>       Case\n",
       "<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">1948</span>       Case\n",
       "Name: batch, Length: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">1949</span>, dtype: category\n",
       "Categories <span style=\"font-weight: bold\">(</span><span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2</span>, object<span style=\"font-weight: bold\">)</span>: <span style=\"font-weight: bold\">[</span><span style=\"color: #008000; text-decoration-color: #008000\">'Case'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Control'</span><span style=\"font-weight: bold\">]</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;36m0\u001b[0m          Case\n",
       "\u001b[1;36m1\u001b[0m          Case\n",
       "\u001b[1;36m2\u001b[0m       Control\n",
       "\u001b[1;36m3\u001b[0m       Control\n",
       "\u001b[1;36m4\u001b[0m          Case\n",
       "         \u001b[33m...\u001b[0m   \n",
       "\u001b[1;36m1944\u001b[0m       Case\n",
       "\u001b[1;36m1945\u001b[0m       Case\n",
       "\u001b[1;36m1946\u001b[0m       Case\n",
       "\u001b[1;36m1947\u001b[0m       Case\n",
       "\u001b[1;36m1948\u001b[0m       Case\n",
       "Name: batch, Length: \u001b[1;36m1949\u001b[0m, dtype: category\n",
       "Categories \u001b[1m(\u001b[0m\u001b[1;36m2\u001b[0m, object\u001b[1m)\u001b[0m: \u001b[1m[\u001b[0m\u001b[32m'Case'\u001b[0m, \u001b[32m'Control'\u001b[0m\u001b[1m]\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: You’re trying to run this on 1199 dimensions of `.X`, if you really want this, set `use_rep='X'`.\n",
      "         Falling back to preprocessing with `sc.pp.pca` and default params.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Found <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">0</span> perturbation markers for <span style=\"font-weight: bold\">(</span><span style=\"color: #008000; text-decoration-color: #008000\">'all'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Case'</span><span style=\"font-weight: bold\">)</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Found \u001b[1;36m0\u001b[0m perturbation markers for \u001b[1m(\u001b[0m\u001b[32m'all'\u001b[0m, \u001b[32m'Case'\u001b[0m\u001b[1m)\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "local_adata = adata.copy()\n",
    "#local_adata.X = local_adata.layers['counts'].astype(np.float32)\n",
    "print(local_adata.obs['batch'])\n",
    "mixscape_identifier = Mixscape()\n",
    "mixscape_identifier.perturbation_signature(adata=local_adata, pert_key='batch', control='Control')\n",
    "mixscape_identifier.mixscape(adata=local_adata, labels='batch', control='Control')\n",
    "# adata.obs['mixscape_class_p_ko'] -> Posterior probabilities used to determine if a cell is KO (default).\n",
    "# adata.obs['mixscape_class_global'] -> Global classification result (perturbed, NP or NT)\n",
    "# adata.obs['mixscape_class'] -> Classification result with cells being either classified as perturbed \n",
    "# (KO, by default) or non-perturbed (NP) based on their target gene class."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "7ff042c7-526d-47e1-aeef-3c11231eacde",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0       Case NP\n",
       "1       Case NP\n",
       "2       Control\n",
       "3       Control\n",
       "4       Case NP\n",
       "         ...   \n",
       "1944    Case NP\n",
       "1945    Case NP\n",
       "1946    Case NP\n",
       "1947    Case NP\n",
       "1948    Case NP\n",
       "Name: mixscape_class, Length: 1949, dtype: object"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "local_adata.obs['mixscape_class']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "916bb284-548f-4675-ab02-adfc4ff65331",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      0.000\n",
       "1      0.000\n",
       "2      0.000\n",
       "3      0.000\n",
       "4      0.000\n",
       "        ... \n",
       "1944   0.000\n",
       "1945   0.000\n",
       "1946   0.000\n",
       "1947   0.000\n",
       "1948   0.000\n",
       "Name: mixscape_class_p_ko, Length: 1949, dtype: float64"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_hat = local_adata.obs['mixscape_class_p_ko'].astype(np.float)\n",
    "\n",
    "p_hat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "bb3064d5-2914-412e-a44d-61dc04e773e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Axes: xlabel='mixscape_class_p_ko', ylabel='Density'>"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjMAAAGxCAYAAACXwjeMAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/bCgiHAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAowUlEQVR4nO3de1TVdb7/8ddO5KICeUm2JiompnlPT4aXsFQMZ4y0ySYdvDSVZlocdUwzzcYVaueI2pie40xL85iXqbTTdEaDScUL3oWsNMcxRMckMhFIDRQ+vz/6uY/EHYG9P5znY629Vvt72b75LItn331zGGOMAAAALHWbuwcAAAC4FcQMAACwGjEDAACsRswAAACrETMAAMBqxAwAALAaMQMAAKxGzAAAAKt5uXuA6lZQUKBvvvlG/v7+cjgc7h4HAACUgzFGOTk5at68uW67rfRrL7U+Zr755hsFBwe7ewwAAFAJZ8+eVYsWLUo9ptbHjL+/v6SfFiMgIMDN0wAAgPLIzs5WcHCw6/d4aWp9zNx4aikgIICYAQDAMuV5iQgvAAYAAFYjZgAAgNWIGQAAYDViBgAAWI2YAQAAViNmAACA1YgZAABgNWIGAABYjZgBAABWI2YAAIDViBkAAGA1YgYAAFiNmAEAAFYjZgAAgNWIGQAAYDUvdw8AALdi3f4z5TpuZK+W1TwJAHfhygwAALAaMQMAAKxGzAAAAKsRMwAAwGrEDAAAsBoxAwAArEbMAAAAqxEzAADAasQMAACwGjEDAACsRswAAACrETMAAMBqxAwAALAaMQMAAKxGzAAAAKsRMwAAwGrEDAAAsBoxAwAArEbMAAAAqxEzAADAasQMAACwGjEDAACsRswAAACrETMAAMBqxAwAALAaMQMAAKzmMTEzf/58ORwOxcTEuLYZYzR37lw1b95cfn5+6t+/v7788kv3DQkAADyOR8TMwYMHtXLlSnXp0qXQ9jfeeENxcXFatmyZDh48KKfTqUGDBiknJ8dNkwIAAE/j9pj54YcfNGrUKP3xj39Uw4YNXduNMVqyZIlmzZql4cOHq1OnTnrnnXd05coVrVu3zo0TAwAAT+L2mHn++ef1i1/8QgMHDiy0PTU1Venp6YqIiHBt8/HxUXh4uJKSkmp6TAAA4KG83PmHb9iwQUeOHNHBgweL7EtPT5ckBQUFFdoeFBSktLS0Eh8zNzdXubm5rvvZ2dlVNC0AAPBEbrsyc/bsWb344otau3atfH19SzzO4XAUum+MKbLtZvPnz1dgYKDrFhwcXGUzAwAAz+O2mDl8+LAyMjLUo0cPeXl5ycvLS4mJiXrzzTfl5eXluiJz4wrNDRkZGUWu1txs5syZysrKct3Onj1brT8HAABwL7c9zTRgwAB9/vnnhbaNGzdO7du310svvaQ2bdrI6XQqISFB3bt3lyTl5eUpMTFRCxcuLPFxfXx85OPjU62zAwAAz+G2mPH391enTp0Kbatfv74aN27s2h4TE6PY2FiFhoYqNDRUsbGxqlevnkaOHOmOkQEAgAdy6wuAyzJ9+nRdvXpVEydOVGZmpnr16qX4+Hj5+/u7ezQAAOAhHMYY4+4hqlN2drYCAwOVlZWlgIAAd48DoIqt23+mXMeN7NWymicBUJUq8vvb7Z8zAwAAcCuIGQAAYDViBgAAWI2YAQAAViNmAACA1YgZAABgNWIGAABYjZgBAABWI2YAAIDViBkAAGA1YgYAAFiNmAEAAFYjZgAAgNWIGQAAYDViBgAAWI2YAQAAViNmAACA1YgZAABgNWIGAABYjZgBAABWI2YAAIDViBkAAGA1YgYAAFiNmAEAAFYjZgAAgNWIGQAAYDViBgAAWI2YAQAAViNmAACA1YgZAABgNWIGAABYjZgBAABWI2YAAIDViBkAAGA1YgYAAFiNmAEAAFYjZgAAgNWIGQAAYDViBgAAWI2YAQAAViNmAACA1YgZAABgNWIGAABYjZgBAABWI2YAAIDViBkAAGA1YgYAAFiNmAEAAFYjZgAAgNWIGQAAYDViBgAAWI2YAQAAViNmAACA1YgZAABgNWIGAABYjZgBAABWI2YAAIDViBkAAGA1YgYAAFiNmAEAAFYjZgAAgNWIGQAAYDViBgAAWI2YAQAAViNmAACA1YgZAABgNWIGAABYjZgBAABWI2YAAIDV3BozK1asUJcuXRQQEKCAgACFhYVpy5Ytrv3GGM2dO1fNmzeXn5+f+vfvry+//NKNEwMAAE/j1php0aKFFixYoEOHDunQoUN66KGHFBUV5QqWN954Q3FxcVq2bJkOHjwop9OpQYMGKScnx51jAwAAD+LWmBk6dKiGDBmidu3aqV27dnr99dfVoEED7du3T8YYLVmyRLNmzdLw4cPVqVMnvfPOO7py5YrWrVvnzrEBAIAH8ZjXzOTn52vDhg26fPmywsLClJqaqvT0dEVERLiO8fHxUXh4uJKSktw4KQAA8CRe7h7g888/V1hYmH788Uc1aNBAmzdv1j333OMKlqCgoELHBwUFKS0trcTHy83NVW5urut+dnZ29QwOAAA8gtuvzNx9991KSUnRvn379Nxzz2nMmDE6duyYa7/D4Sh0vDGmyLabzZ8/X4GBga5bcHBwtc0OAADcz+0x4+3trbZt26pnz56aP3++unbtqqVLl8rpdEqS0tPTCx2fkZFR5GrNzWbOnKmsrCzX7ezZs9U6PwAAcC+3x8zPGWOUm5urkJAQOZ1OJSQkuPbl5eUpMTFRvXv3LvF8Hx8f11u9b9wAAEDt5dbXzLz88suKjIxUcHCwcnJytGHDBu3YsUNbt26Vw+FQTEyMYmNjFRoaqtDQUMXGxqpevXoaOXKkO8cGAAAexK0x8+233yo6Olrnz59XYGCgunTpoq1bt2rQoEGSpOnTp+vq1auaOHGiMjMz1atXL8XHx8vf39+dYwMAAA/iMMYYdw9RnbKzsxUYGKisrCyecgJqoXX7z5TruJG9WlbzJACqUkV+f3vca2YAAAAqgpgBAABWI2YAAIDViBkAAGA1YgYAAFiNmAEAAFYjZgAAgNWIGQAAYDViBgAAWI2YAQAAViNmAACA1YgZAABgNWIGAABYjZgBAABWI2YAAIDViBkAAGA1YgYAAFiNmAEAAFYjZgAAgNWIGQAAYDViBgAAWI2YAQAAViNmAACA1YgZAABgNWIGAABYrVIxk5qaWtVzAAAAVEqlYqZt27Z68MEHtXbtWv34449VPRMAAEC5VSpmPvvsM3Xv3l1Tp06V0+nU+PHjdeDAgaqeDQAAoEyViplOnTopLi5O586d06pVq5Senq6+ffuqY8eOiouL03fffVfVcwIAABTrll4A7OXlpWHDhunPf/6zFi5cqFOnTmnatGlq0aKFRo8erfPnz1fVnAAAAMW6pZg5dOiQJk6cqGbNmikuLk7Tpk3TqVOntG3bNp07d05RUVFVNScAAECxvCpzUlxcnFatWqUTJ05oyJAhWrNmjYYMGaLbbvupjUJCQvSf//mfat++fZUOCwAA8HOVipkVK1boqaee0rhx4+R0Oos9pmXLlnr77bdvaTgAAICyVCpmEhIS1LJlS9eVmBuMMTp79qxatmwpb29vjRkzpkqGBAAAKEmlXjNz11136cKFC0W2X7x4USEhIbc8FAAAQHlVKmaMMcVu/+GHH+Tr63tLAwEAAFREhZ5mmjJliiTJ4XBozpw5qlevnmtffn6+9u/fr27dulXpgAAAAKWpUMwkJydL+unKzOeffy5vb2/XPm9vb3Xt2lXTpk2r2gkBAABKUaGY2b59uyRp3LhxWrp0qQICAqplKAAAgPKq1LuZVq1aVdVzAAAAVEq5Y2b48OFavXq1AgICNHz48FKP3bRp0y0PBgAAUB7ljpnAwEA5HA7XPwMAAHiCcsfMzU8t8TQTAADwFJX6nJmrV6/qypUrrvtpaWlasmSJ4uPjq2wwAACA8qhUzERFRWnNmjWSpEuXLum+++7TokWLFBUVpRUrVlTpgAAAAKWpVMwcOXJE/fr1kyS9//77cjqdSktL05o1a/Tmm29W6YAAAAClqVTMXLlyRf7+/pKk+Ph4DR8+XLfddpvuv/9+paWlVemAAAAApalUzLRt21Yffvihzp49q08++UQRERGSpIyMDD5IDwAA1KhKxcycOXM0bdo0tW7dWr169VJYWJikn67SdO/evUoHBAAAKE2lPgH4V7/6lfr27avz58+ra9euru0DBgzQsGHDqmw4AACAslQqZiTJ6XTK6XQW2nbffffd8kAAAAAVUamYuXz5shYsWKBPP/1UGRkZKigoKLT/66+/rpLhAAAAylKpmHn66aeVmJio6OhoNWvWzPU1BwAAADWtUjGzZcsW/c///I/69OlT1fMAAABUSKXezdSwYUM1atSoqmcBAACosErFzLx58zRnzpxC388EAADgDpV6mmnRokU6deqUgoKC1Lp1a9WtW7fQ/iNHjlTJcAAAAGWpVMw8+uijVTwGAABA5VQqZl599dWqngMAAKBSKvWaGUm6dOmS/vSnP2nmzJm6ePGipJ+eXjp37lyVDQcAAFCWSl2ZOXr0qAYOHKjAwECdPn1azzzzjBo1aqTNmzcrLS1Na9asqeo5AQAAilWpKzNTpkzR2LFjdfLkSfn6+rq2R0ZGaufOnVU2HAAAQFkqFTMHDx7U+PHji2y/8847lZ6efstDAQAAlFelYsbX11fZ2dlFtp84cUJ33HHHLQ8FAABQXpWKmaioKP3+97/XtWvXJEkOh0NnzpzRjBkz9Nhjj1XpgAAAAKWpVMz8+7//u7777js1bdpUV69eVXh4uNq2bSt/f3+9/vrrVT0jAABAiSr1bqaAgADt3r1b27dv1+HDh1VQUKB7771XAwcOrOr5AAAASlXhmCkoKNDq1au1adMmnT59Wg6HQyEhIXI6nTLGyOFwVMecAAAAxarQ00zGGD3yyCN6+umnde7cOXXu3FkdO3ZUWlqaxo4dq2HDhlXXnAAAAMWqUMysXr1aO3fu1Keffqrk5GStX79eGzZs0Geffaa//e1v2rZtW4U+MG/+/Pn6l3/5F/n7+6tp06Z69NFHdeLEiULHGGM0d+5cNW/eXH5+furfv7++/PLLiowNAABqsQrFzPr16/Xyyy/rwQcfLLLvoYce0owZM/Tuu++W+/ESExP1/PPPa9++fUpISND169cVERGhy5cvu4554403FBcXp2XLlungwYNyOp0aNGiQcnJyKjI6AACopRzGGFPeg51Op7Zu3apu3boVuz85OVmRkZGV/uC8G++QSkxM1AMPPCBjjJo3b66YmBi99NJLkqTc3FwFBQVp4cKFxX5w389lZ2crMDBQWVlZCggIqNRcADzXuv1nynXcyF4tq3kSAFWpIr+/K3Rl5uLFiwoKCipxf1BQkDIzMyvykIVkZWVJkho1aiRJSk1NVXp6uiIiIlzH+Pj4KDw8XElJSZX+cwAAQO1RoXcz5efny8ur5FPq1Kmj69evV2oQY4ymTJmivn37qlOnTpLkusLz84AKCgpSWlpasY+Tm5ur3Nxc1/3iPqkYAADUHhWKGWOMxo4dKx8fn2L33xwRFTVp0iQdPXpUu3fvLrLv52/3Lu0t4PPnz9drr71W6TkAAIBdKhQzY8aMKfOY0aNHV3iIyZMn66OPPtLOnTvVokUL13an0ynppys0zZo1c23PyMgo8emumTNnasqUKa772dnZCg4OrvBMAADADhWKmVWrVlXpH26M0eTJk7V582bt2LFDISEhhfbf+DC+hIQEde/eXZKUl5enxMRELVy4sNjH9PHxKfHKEQAAqH0q9XUGVeX555/XunXr9N///d/y9/d3vUYmMDBQfn5+cjgciomJUWxsrEJDQxUaGqrY2FjVq1dPI0eOdOfoAADAQ7g1ZlasWCFJ6t+/f6Htq1at0tixYyVJ06dP19WrVzVx4kRlZmaqV69eio+Pl7+/fw1PCwAAPFGFPmfGRnzODFC78TkzQO1UbZ8zAwAA4GmIGQAAYDViBgAAWI2YAQAAViNmAACA1YgZAABgNWIGAABYjZgBAABWI2YAAIDViBkAAGA1YgYAAFiNmAEAAFYjZgAAgNWIGQAAYDViBgAAWI2YAQAAViNmAACA1YgZAABgNWIGAABYjZgBAABWI2YAAIDViBkAAGA1YgYAAFiNmAEAAFYjZgAAgNWIGQAAYDViBgAAWI2YAQAAViNmAACA1YgZAABgNWIGAABYjZgBAABWI2YAAIDViBkAAGA1YgYAAFiNmAEAAFYjZgAAgNWIGQAAYDViBgAAWI2YAQAAViNmAACA1YgZAABgNWIGAABYjZgBAABWI2YAAIDViBkAAGA1YgYAAFiNmAEAAFYjZgAAgNWIGQAAYDViBgAAWI2YAQAAViNmAACA1YgZAABgNWIGAABYjZgBAABWI2YAAIDViBkAAGA1YgYAAFiNmAEAAFYjZgAAgNWIGQAAYDViBgAAWI2YAQAAViNmAACA1YgZAABgNWIGAABYjZgBAABWI2YAAIDV3BozO3fu1NChQ9W8eXM5HA59+OGHhfYbYzR37lw1b95cfn5+6t+/v7788kv3DAsAADySW2Pm8uXL6tq1q5YtW1bs/jfeeENxcXFatmyZDh48KKfTqUGDBiknJ6eGJwUAAJ7Ky51/eGRkpCIjI4vdZ4zRkiVLNGvWLA0fPlyS9M477ygoKEjr1q3T+PHja3JUAADgoTz2NTOpqalKT09XRESEa5uPj4/Cw8OVlJTkxskAAIAnceuVmdKkp6dLkoKCggptDwoKUlpaWonn5ebmKjc313U/Ozu7egYEAAAewWOvzNzgcDgK3TfGFNl2s/nz5yswMNB1Cw4Oru4RAQCAG3lszDidTkn/e4XmhoyMjCJXa242c+ZMZWVluW5nz56t1jkBAIB7eWzMhISEyOl0KiEhwbUtLy9PiYmJ6t27d4nn+fj4KCAgoNANAADUXm59zcwPP/ygf/zjH677qampSklJUaNGjdSyZUvFxMQoNjZWoaGhCg0NVWxsrOrVq6eRI0e6cWoAAOBJ3Bozhw4d0oMPPui6P2XKFEnSmDFjtHr1ak2fPl1Xr17VxIkTlZmZqV69eik+Pl7+/v7uGhkAAHgYhzHGuHuI6pSdna3AwEBlZWXxlBNQC63bf6Zcx43s1bKaJwFQlSry+9tjXzMDAABQHsQMAACwGjEDAACsRswAAACrETMAAMBqxAwAALAaMQMAAKxGzAAAAKsRMwAAwGrEDAAAsBoxAwAArEbMAAAAqxEzAADAasQMAACwGjEDAACsRswAAACrETMAAMBqxAwAALAaMQMAAKxGzAAAAKsRMwAAwGrEDAAAsBoxAwAArEbMAAAAqxEzAADAasQMAACwGjEDAACsRswAAACrETMAAMBqxAwAALAaMQMAAKxGzAAAAKsRMwAAwGrEDAAAsBoxAwAArEbMAAAAqxEzAADAasQMAACwGjEDAACsRswAAACrETMAAMBqxAwAALAaMQMAAKxGzAAAAKsRMwAAwGrEDAAAsBoxAwAArEbMAAAAqxEzAADAasQMAACwGjEDAACsRswAAACrETMAAMBqxAwAALAaMQMAAKxGzAAAAKsRMwAAwGrEDAAAsBoxAwAArEbMAAAAqxEzAADAasQMAACwGjEDAACsRswAAACrETMAAMBqxAwAALAaMQMAAKxGzAAAAKsRMwAAwGpWxMzy5csVEhIiX19f9ejRQ7t27XL3SAAAwEN4fMxs3LhRMTExmjVrlpKTk9WvXz9FRkbqzJkz7h4NAAB4AI+Pmbi4OP32t7/V008/rQ4dOmjJkiUKDg7WihUr3D0aAADwAB4dM3l5eTp8+LAiIiIKbY+IiFBSUpKbpgIAAJ7Ey90DlObChQvKz89XUFBQoe1BQUFKT08v9pzc3Fzl5ua67mdlZUmSsrOzq29QAG5z5XJOuY7jvwGAXW78O2uMKfNYj46ZGxwOR6H7xpgi226YP3++XnvttSLbg4ODq2U2AHZ4xt0DAKiUnJwcBQYGlnqMR8dMkyZNVKdOnSJXYTIyMopcrblh5syZmjJliut+QUGBLl68qMaNG5cYQP+XZGdnKzg4WGfPnlVAQIC7x6m1WOeawTrXDNa5ZrDOhRljlJOTo+bNm5d5rEfHjLe3t3r06KGEhAQNGzbMtT0hIUFRUVHFnuPj4yMfH59C226//fbqHNNKAQEB/MtSA1jnmsE61wzWuWawzv+rrCsyN3h0zEjSlClTFB0drZ49eyosLEwrV67UmTNnNGHCBHePBgAAPIDHx8wTTzyh77//Xr///e91/vx5derUSX/961/VqlUrd48GAAA8gMfHjCRNnDhREydOdPcYtYKPj49effXVIk/FoWqxzjWDda4ZrHPNYJ0rz2HK854nAAAAD+XRH5oHAABQFmIGAABYjZgBAABWI2ZquczMTEVHRyswMFCBgYGKjo7WpUuXyn3++PHj5XA4tGTJkmqbsTao6Dpfu3ZNL730kjp37qz69eurefPmGj16tL755puaG9oSy5cvV0hIiHx9fdWjRw/t2rWr1OMTExPVo0cP+fr6qk2bNvqP//iPGprUbhVZ502bNmnQoEG64447FBAQoLCwMH3yySc1OK29Kvr3+YY9e/bIy8tL3bp1q94BLUXM1HIjR45USkqKtm7dqq1btyolJUXR0dHlOvfDDz/U/v37y/Xpi//XVXSdr1y5oiNHjmj27Nk6cuSINm3apL///e965JFHanBqz7dx40bFxMRo1qxZSk5OVr9+/RQZGakzZ84Ue3xqaqqGDBmifv36KTk5WS+//LJeeOEFffDBBzU8uV0qus47d+7UoEGD9Ne//lWHDx/Wgw8+qKFDhyo5ObmGJ7dLRdf5hqysLI0ePVoDBgyooUktZFBrHTt2zEgy+/btc23bu3evkWS++uqrUs/95z//ae68807zxRdfmFatWpnFixdX87T2upV1vtmBAweMJJOWllYdY1rpvvvuMxMmTCi0rX379mbGjBnFHj99+nTTvn37QtvGjx9v7r///mqbsTao6DoX55577jGvvfZaVY9Wq1R2nZ944gnzyiuvmFdffdV07dq1Gie0F1dmarG9e/cqMDBQvXr1cm27//77FRgYqKSkpBLPKygoUHR0tH73u9+pY8eONTGq1Sq7zj+XlZUlh8PB12/8f3l5eTp8+LAiIiIKbY+IiChxXffu3Vvk+MGDB+vQoUO6du1atc1qs8qs888VFBQoJydHjRo1qo4Ra4XKrvOqVat06tQpvfrqq9U9otWs+NA8VE56erqaNm1aZHvTpk2LfHnnzRYuXCgvLy+98MIL1TlerVHZdb7Zjz/+qBkzZmjkyJF8J8v/d+HCBeXn5xf5UtmgoKAS1zU9Pb3Y469fv64LFy6oWbNm1TavrSqzzj+3aNEiXb58WSNGjKiOEWuFyqzzyZMnNWPGDO3atUteXvy6Lg1XZiw0d+5cORyOUm+HDh2SpGK/KdwYU+I3iB8+fFhLly7V6tWr/89/y3h1rvPNrl27pl//+tcqKCjQ8uXLq/znsN3P17CsdS3u+OK2o7CKrvMN69ev19y5c7Vx48Ziox6FlXed8/PzNXLkSL322mtq165dTY1nLVLPQpMmTdKvf/3rUo9p3bq1jh49qm+//bbIvu+++67I/x3csGvXLmVkZKhly5aubfn5+Zo6daqWLFmi06dP39LsNqnOdb7h2rVrGjFihFJTU7Vt2zauytykSZMmqlOnTpH/a83IyChxXZ1OZ7HHe3l5qXHjxtU2q80qs843bNy4Ub/97W/13nvvaeDAgdU5pvUqus45OTk6dOiQkpOTNWnSJEk/PZ1njJGXl5fi4+P10EMP1cjsNiBmLNSkSRM1adKkzOPCwsKUlZWlAwcO6L777pMk7d+/X1lZWerdu3ex50RHRxf5j9LgwYMVHR2tcePG3frwFqnOdZb+N2ROnjyp7du388v2Z7y9vdWjRw8lJCRo2LBhru0JCQmKiooq9pywsDD95S9/KbQtPj5ePXv2VN26dat1XltVZp2ln67IPPXUU1q/fr1+8Ytf1MSoVqvoOgcEBOjzzz8vtG358uXatm2b3n//fYWEhFT7zFZx44uPUQMefvhh06VLF7N3716zd+9e07lzZ/PLX/6y0DF333232bRpU4mPwbuZylbRdb527Zp55JFHTIsWLUxKSoo5f/6865abm+uOH8EjbdiwwdStW9e8/fbb5tixYyYmJsbUr1/fnD592hhjzIwZM0x0dLTr+K+//trUq1fP/Ou//qs5duyYefvtt03dunXN+++/764fwQoVXed169YZLy8v89ZbbxX6u3vp0iV3/QhWqOg6/xzvZioZMVPLff/992bUqFHG39/f+Pv7m1GjRpnMzMxCx0gyq1atKvExiJmyVXSdU1NTjaRib9u3b6/x+T3ZW2+9ZVq1amW8vb3NvffeaxITE137xowZY8LDwwsdv2PHDtO9e3fj7e1tWrdubVasWFHDE9upIuscHh5e7N/dMWPG1Pzglqno3+ebETMl41uzAQCA1Xg3EwAAsBoxAwAArEbMAAAAqxEzAADAasQMAACwGjEDAACsRswAAACrETMAAMBqxAxQy+zYsUMOh0OXLl1y9ygeoX///oqJiXH3GLfk9OnTcjgcSklJcfcogEciZoBapnfv3jp//rwCAwPdPQoA1Ai+NRuoZby9veV0Ot09BgDUGK7MAB6uf//+mjx5smJiYtSwYUMFBQVp5cqVunz5ssaNGyd/f3/ddddd2rJli6SiTzM99dRT6tKli3JzcyVJ165dU48ePTRq1ChJUl5eniZNmqRmzZrJ19dXrVu31vz5811//qVLl/Tss88qKChIvr6+6tSpkz7++GNJ0vfff68nn3xSLVq0UL169dS5c2etX7++yPyTJk3SpEmTdPvtt6tx48Z65ZVXdPPXwuXl5Wn69Om68847Vb9+ffXq1Us7duwo9xrt2bNH4eHhqlevnho2bKjBgwcrMzOz2GPXrl2rnj17yt/fX06nUyNHjlRGRoZrf2ZmpkaNGqU77rhDfn5+Cg0N1apVq8q1VqVxOBxasWKFIiMj5efnp5CQEL333nvl/hlvVlBQoGeeeUbt2rVTWlqaJGnFihW666675O3trbvvvlv/9V//VanHBmxEzAAWeOedd9SkSRMdOHBAkydP1nPPPafHH39cvXv31pEjRzR48GBFR0frypUrRc598803dfnyZc2YMUOSNHv2bF24cEHLly937f/oo4/05z//WSdOnNDatWvVunVrST/90oyMjFRSUpLWrl2rY8eOacGCBapTp44k6ccff1SPHj308ccf64svvtCzzz6r6Oho7d+/v8j8Xl5e2r9/v958800tXrxYf/rTn1z7x40bpz179mjDhg06evSoHn/8cT388MM6efJkmWuTkpKiAQMGqGPHjtq7d692796toUOHKj8/v9jj8/LyNG/ePH322Wf68MMPlZqaqrFjx7r2z549W8eOHdOWLVt0/PhxrVixQk2aNClzrcpj9uzZeuyxx/TZZ5/pN7/5jZ588kkdP3683OffmH/EiBE6dOiQdu/erVatWmnz5s168cUXNXXqVH3xxRcaP368xo0bp+3bt1fosQFruflbuwGUITw83PTt29d1//r166Z+/fomOjrate38+fNGktm7d6/Zvn27kWQyMzNd+5OSkkzdunXN7NmzjZeXl0lMTHTtmzx5snnooYdMQUFBkT/7k08+Mbfddps5ceJEuecdMmSImTp1aqH5O3ToUOjxX3rpJdOhQwdjjDH/+Mc/jMPhMOfOnSv0OAMGDDAzZ84s88978sknTZ8+fUrcHx4ebl588cUS9x84cMBIMjk5OcYYY4YOHWrGjRtX7LGlrVVZJJkJEyYU2tarVy/z3HPPlXluamqqkWR27dplBg4caPr06WMuXbrk2t+7d2/zzDPPFDrn8ccfN0OGDKnwnICNuDIDWKBLly6uf65Tp44aN26szp07u7YFBQVJUqGnS24WFhamadOmad68eZo6daoeeOAB176xY8cqJSVFd999t1544QXFx8e79qWkpKhFixZq165dsY+bn5+v119/XV26dFHjxo3VoEEDxcfH68yZM4WOu//+++VwOArNc/LkSeXn5+vIkSMyxqhdu3Zq0KCB65aYmKhTp06VuTY3rsyUV3JysqKiotSqVSv5+/urf//+kuSa+bnnntOGDRvUrVs3TZ8+XUlJSeVaq/IICwsrcr8iV2aefPJJ/fDDD4qPjy/0Au/jx4+rT58+hY7t06dPha/6ALYiZgAL1K1bt9B9h8NRaNuNUCgoKCj2/IKCAu3Zs0d16tQp8tTNvffeq9TUVM2bN09Xr17ViBEj9Ktf/UqS5OfnV+pcixYt0uLFizV9+nRt27ZNKSkpGjx4sPLy8sr9sxUUFKhOnTo6fPiwUlJSXLfjx49r6dKlZZ5f1ow3u3z5siIiItSgQQOtXbtWBw8e1ObNmyXJNXNkZKTS0tIUExOjb775RgMGDNC0adMklb5WlXVz5JVlyJAhOnr0qPbt21fm4xhjKvTYgM2IGeD/gH/7t3/T8ePHlZiYqE8++cT1gtYbAgIC9MQTT+iPf/yjNm7cqA8++EAXL15Uly5d9M9//lN///vfi33cXbt2KSoqSr/5zW/UtWtXtWnTptjXufz8l+++ffsUGhqqOnXqqHv37srPz1dGRobatm1b6Faed2V16dJFn376abnW4auvvtKFCxe0YMEC9evXT+3bty/2atYdd9yhsWPHau3atVqyZIlWrlzp2lfSWpVHcevQvn37cp0r/XTVaMGCBXrkkUeUmJjo2t6hQwft3r270LFJSUnq0KFDuR8bsBlvzQZquZSUFM2ZM0fvv/+++vTpo6VLl+rFF19UeHi42rRpo8WLF6tZs2bq1q2bbrvtNr333ntyOp26/fbbFR4ergceeECPPfaY4uLi1LZtW3311VdyOBx6+OGH1bZtW33wwQdKSkpSw4YNFRcXp/T09CK/RM+ePaspU6Zo/PjxOnLkiP7whz9o0aJFkqR27dpp1KhRGj16tBYtWqTu3bvrwoUL2rZtmzp37qwhQ4aU+vPNnDlTnTt31sSJEzVhwgR5e3tr+/btevzxx10v3L2hZcuW8vb21h/+8AdNmDBBX3zxhebNm1fomDlz5qhHjx7q2LGjcnNz9fHHH7t+ntLWqjzee+899ezZU3379tW7776rAwcO6O233y7XuTdMnjxZ+fn5+uUvf6ktW7aob9+++t3vfqcRI0bo3nvv1YABA/SXv/xFmzZt0t/+9rcKPTZgLXe/aAdA6Yp7AWurVq3M4sWLC22TZDZv3lzoBcBXr14199xzj3n22WcLHTts2DDTu3dvc/36dbNy5UrTrVs3U79+fRMQEGAGDBhgjhw54jr2+++/N+PGjTONGzc2vr6+plOnTubjjz927YuKijINGjQwTZs2Na+88ooZPXq0iYqKKjT/xIkTzYQJE0xAQIBp2LChmTFjRqEX0ebl5Zk5c+aY1q1bm7p16xqn02mGDRtmjh49Wq412rFjh+ndu7fx8fExt99+uxk8eLDrBdA/X79169aZ1q1bGx8fHxMWFmY++ugjI8kkJycbY4yZN2+e6dChg/Hz8zONGjUyUVFR5uuvvzbGmDLXqjSSzFtvvWUGDRpkfHx8TKtWrcz69evLde6NFwDfmNEYYxYtWmT8/f3Nnj17jDHGLF++3LRp08bUrVvXtGvXzqxZs6Zcjw3UBg5jbvqwBwCoYv3791e3bt20ZMkSd4/iVg6HQ5s3b9ajjz7q7lGAWofXzAAAAKsRMwA8WmRkZKG3bN98i42Ndfd4kqR33323xBk7duxY5vmxsbElnh8ZGVkDPwFgN55mAuDRzp07p6tXrxa7r1GjRmrUqFENT1RUTk6Ovv3222L31a1bV61atSr1/IsXL5b4jig/Pz/deeedtzwjUJsRMwAAwGo8zQQAAKxGzAAAAKsRMwAAwGrEDAAAsBoxAwAArEbMAAAAqxEzAADAasQMAACw2v8DqVvcrcwn51MAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.distplot(p_hat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c758925-306c-4ca6-a149-11334fb0cee7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "environment": {
   "name": "pytorch-gpu.1-9.m81",
   "type": "gcloud",
   "uri": "gcr.io/deeplearning-platform-release/pytorch-gpu.1-9:m81"
  },
  "kernelspec": {
   "display_name": "Python [conda env:hidden]",
   "language": "python",
   "name": "conda-env-hidden-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
